prow/config/branch_protection.go

```diff
@@ -22,102 +22,123 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 )
 
+type PolicyStruct struct {
+	Protect *bool `json:"protect-by-default,omitempty"`
+	// TODO(fejta): add all protection options
+	Contexts []string `json:"require-contexts,omitempty"`
+	Pushers  []string `json:"allow-push,omitempty"`
+}
+
+// selectBool returns the child argument if set, otherwise the parent
+func selectBool(parent, child *bool) *bool {
+	if child != nil {
+		return child
+	}
+	return parent
+}
+
+// unionStrings merges the parent and child items together
+func unionStrings(parent, child []string) []string {
+	if child == nil {
+		return parent
+	}
+	if parent == nil {
+		return child
+	}
+	s := sets.NewString(parent...)
+	s.Insert(child...)
+	return s.List()
+}
+
+// apply returns a policy that merges the child into the parent
+func (parent PolicyStruct) Apply(child Policy) Policy {
+	p := parent.Get()
+	c := child.Get()
+	return PolicyStruct{
+		Protect:  selectBool(p.Protect, c.Protect),
+		Contexts: unionStrings(p.Contexts, c.Contexts),
+		Pushers:  unionStrings(p.Pushers, c.Pushers),
+	}
+}
+
+func (p PolicyStruct) Get() PolicyStruct {
+	return p
+}
+
+type Policy interface {
```

Comment:
What is the purpose of this interface? Since `Get` always returns `PolicyStruct` it seems like this interface doesn't provide any advantages over simply using a `PolicyStruct` everywhere. I think the interface and all of the `Get` calls could be removed.

Reply:
Done, do you like this better?